// node-jvm/libs/classfile/classarea.js
/*
 node-jvm
 Copyright (c) 2013 Yaroslav Gaponov <yaroslav.gaponov@gmail.com>
*/

var util = require("util"),
    Reader = require("../util/reader.js"),
    TAGS = require("./tags.js"),
    ACCESS_FLAGS = require("./accessflags.js"),
    ATTRIBUTE_TYPES = require("./attributetypes.js");


var ClassArea = module.exports = function(classBytes) {
    if (this instanceof ClassArea) {
        this.classImage = getClassImage(classBytes);
    } else {
        return new ClassArea(classBytes);
    }
}

ClassArea.prototype.getClassName = function() {    
    return this.classImage.constant_pool[this.classImage.constant_pool[this.classImage.this_class].name_index].bytes;    
}

ClassArea.prototype.getSuperClassName = function() {    
    return this.classImage.constant_pool[this.classImage.constant_pool[this.classImage.super_class].name_index].bytes;    
}

ClassArea.prototype.getAccessFlags = function() {
    return this.classImage.access_flags;    
}

ClassArea.prototype.getConstantPool = function() {
    return this.classImage.constant_pool;
}

ClassArea.prototype.getFields = function() {
    return this.classImage.fields;
}

ClassArea.prototype.getMethods = function() {
    return this.classImage.methods;
}

ClassArea.prototype.getClasses = function() {
    var self = this;
    var classes = [];
    this.classImage.attributes.forEach(function(a) {
        if (a.info.type === ATTRIBUTE_TYPES.InnerClasses) {
            a.info.classes.forEach(function(c) {
                classes.push(self.classImage.constant_pool[self.classImage.constant_pool[c.inner_class_info_index].name_index].bytes);
                classes.push(self.classImage.constant_pool[self.classImage.constant_pool[c.outer_class_info_index].name_index].bytes);
            });
        }
    });
    return classes;
}

var getClassImage = function(classBytes) {
    //console.log(classBytes)
    var classImage = {};
        
    var getAttribues = function(attribute_name_index, bytes) {
            
        var reader = new Reader.create(bytes);
        var attribute = { attribute_name_index: attribute_name_index };
    
    
        var item = classImage.constant_pool[attribute_name_index];
    
        
        switch(item.tag) {
        
            case TAGS.CONSTANT_Long:
            case TAGS.CONSTANT_Float:
            case TAGS.CONSTANT_Double:
            case TAGS.CONSTANT_Integer:
            case TAGS.CONSTANT_String:
                attribute.type = ATTRIBUTE_TYPES.ConstantValue;
                attribute.constantvalue_index = reader.read16();
                return attribute;         
                
            case TAGS.CONSTANT_Utf8:
                switch(item.bytes) {
                    
                    case ATTRIBUTE_TYPES.Code:
                        attribute.type = ATTRIBUTE_TYPES.Code;
                        attribute.max_stack = reader.read16();
                        attribute.max_locals = reader.read16();
                        var code_length = reader.read32();
                        attribute.code = reader.readBytes(code_length);
                        
                        var exception_table_length = reader.read16();        
                        attribute.exception_table = [];
                        for(var i=0; i<exception_table_length; i++) {
                            var start_pc = reader.read16();
                            var end_pc = reader.read16();
                            var handler_pc= reader.read16();
                            var catch_type = reader.read16();
                            attribute.exception_table.push({start_pc:start_pc,end_pc:end_pc,handler_pc:handler_pc,catch_type:catch_type });
                        }

                        var attributes_count = reader.read16();        
                        attribute.attributes = [];
                        for(var i=0; i<attributes_count; i++) {
                            var attribute_name_index = reader.read16();
                            var attribute_length = reader.read32();
                            var info = reader.readBytes(attribute_length);
                            attribute.attributes.push({ attribute_name_index: attribute_name_index, attribute_length: attribute_length, info: info });
                        }
                        return attribute;
                        
                    case ATTRIBUTE_TYPES.SourceFile:
                        attribute.type = ATTRIBUTE_TYPES.SourceFile;
                        attribute.sourcefile_index = reader.read16();
                        return attribute;
                    
                    case ATTRIBUTE_TYPES.Exceptions:
                        attribute.type = ATTRIBUTE_TYPES.Exceptions;
                        var number_of_exceptions = reader.read16();
                        attribute.exception_index_table = [];
                        for(var i=0; i<number_of_exceptions; i++) {
                            attribute.exception_index_table.push(reader.read16());
                        }
                        return attribute;
                    
                    case ATTRIBUTE_TYPES.InnerClasses:
                        attribute.type = ATTRIBUTE_TYPES.InnerClasses;
                        var number_of_classes = reader.read16();
                        attribute.classes = [];
                        for(var i=0; i<number_of_classes; i++) {
                            var inner = {};
                            inner.inner_class_info_index = reader.read16();
                            inner.outer_class_info_index = reader.read16();
                            inner.inner_name_index = reader.read16();
                            inner.inner_class_access_flags = reader.read16();
                            attribute.classes.push(inner);
                        }
                        return attribute;
                    // case 'EnclosingMethod':
                    //     var attribute_name_index= reader.read16();
                    //     var attribute_length= reader.read32();
                    //     var class_index = reader.read16();
                    //     var method_index= reader.read16();
                    //     return attribute;
                    // case 'Signature':
                    //     var attribute_name_index= reader.read16(); //指向常量池的"Signature"
                    //     var attribute_length= reader.read32(); // 为2
                    //     var signature_index= reader.read16();
                    //     return attribute;
                    case 'BootstrapMethods':
                        // BootStrapMethods_attribute {
                        //     u2 attribute_name_index; // 属性名称，指向常量池索引，这里为 “BootstrapMethods”
                        //     u4 attribute_length; // 4字节数字，表示属性长度（不含前六个字节）
                        //     u2 num_bootstrap_methods; // 表示引导方法的数量
                        //     {
                        //         u2 bootstrap_method_ref; // 指向常量池的常数，对应的常量池项类型为 CONSTANT_MethodHandle，用于指明实际的引导函数
                        //         u2 num_bootstrap_arguments; // 指明引导函数的参数个数
                        //         u2 bootstrap_arguments[num_bootstrap_arguments]; // 参数的内容
                        //     } bootstrap_methods[num_bootstrap_methods]; // num_bootstrap_methods 个引导方法
                        // }
                        //console.log(reader.bytes) 
                        reader.readBytes(reader.bytes.length);
                        // var attribute_name_index= reader.read16();
                        // var attribute_length= reader.read32(); 
                        // var num_bootstrap_methods = reader.read16();
                        // console.log(num_bootstrap_methods)
                        // for(var i=0;i<num_bootstrap_methods;i++)
                        // { 
                        //     var bootstrap_method_ref=reader.read16();
                        //     var num_bootstrap_arguments=reader.read16();
                        //     for(var j=0;j<num_bootstrap_arguments;j++)
                        //     {
                        //        var bootstrap_arguments =reader.read16();
                        //     } 
                        // } 
                        return attribute;
                    default:
                        reader.readBytes(reader.bytes.length);
                        //throw new Error("This attribute type is not supported yet. [" + JSON.stringify(item) + "]");        
                       console.warn("This attribute type is not supported yet. [" + JSON.stringify(item) + "]");         
                }
                
            default:
                reader.readBytes(reader.bytes.length);
                console.warn("This attribute type is not supported yet. [" + JSON.stringify(item) + "]");    
                //throw new Error("This attribute type is not supported yet. [" + JSON.stringify(item) + "]");            
        }
    };
    
    //console.log(classBytes)
    var reader = Reader.create(classBytes);
    classImage.magic = reader.read32().toString(16);

    classImage.version = {
        minor_version: reader.read16(),
        major_version: reader.read16()
    };
    
    classImage.constant_pool = [ null ];
    var constant_pool_count = reader.read16();
    console.log(classBytes)
    for(var i=1; i<constant_pool_count; i++) {        
        var tag =  reader.read8();
        //console.log("tag",tag);
        switch(tag) {
            case TAGS.CONSTANT_Class:
                var name_index = reader.read16();
                classImage.constant_pool.push( { tag: tag, name_index: name_index } );
                break;
            case TAGS.CONSTANT_Utf8:
                var length = reader.read16();
                var bytes = reader.readString(length);
                //console.log(bytes)
                classImage.constant_pool.push( { tag: tag, bytes: bytes } );
                break;
            case TAGS.CONSTANT_Methodref:
                var class_index = reader.read16();
                var name_and_type_index = reader.read16();
                classImage.constant_pool.push( {  tag: tag, class_index: class_index, name_and_type_index: name_and_type_index } );
                break;
            case TAGS.CONSTANT_NameAndType:
                var name_index = reader.read16();
                var signature_index = reader.read16();
                classImage.constant_pool.push( { tag: tag, name_index: name_index,  signature_index: signature_index } );                
                break;
            case TAGS.CONSTANT_Fieldref:
                var class_index = reader.read16();
                var name_and_type_index = reader.read16();
                classImage.constant_pool.push( { tag: tag, class_index: class_index,  name_and_type_index: name_and_type_index } );                                
                break;
            case TAGS.CONSTANT_String:
                var string_index = reader.read16();
                classImage.constant_pool.push( { tag: tag, string_index: string_index } );                                                
                break;
            case TAGS.CONSTANT_Float:
            case TAGS.CONSTANT_Integer:
                var bytes = reader.read32();
                classImage.constant_pool.push( {  tag: tag, bytes: bytes } );                                                
                break; 
            case TAGS.CONSTANT_Double:
            case TAGS.CONSTANT_Long:
                var bytes = new Buffer(8);
                for (var b=0; b<8; b++) {
                    bytes[b] = reader.read8();
                }
                classImage.constant_pool.push( {  tag: tag, bytes: bytes } );
                classImage.constant_pool.push( null ); i++;
                break;
            case TAGS.CONSTANT_Fieldref:
            case TAGS.CONSTANT_Methodref:
            case TAGS.CONSTANT_InterfaceMethodref:
                var class_index = reader.read16();
                var name_and_type_index = reader.read16();
                classImage.constant_pool.push( {  tag: tag, class_index: class_index, name_and_type_index:name_and_type_index } );                                                
                break;
            case TAGS.CONSTANT_MethodType:
                var descriptor_index = reader.read16();
                classImage.constant_pool.push( {  tag: tag, descriptor_index: descriptor_index} );    
                break;    
            case TAGS.CONSTANT_MethodHandle:
                var reference_kind = reader.read8();
                var reference_index = reader.read16(); 
                classImage.constant_pool.push( {  tag: tag, reference_kind: reference_kind,reference_index:reference_index} );   
                break;
            case TAGS.CONSTANT_InvokeDynamic:
                var bootstrap_method_attr_index = reader.read16();
                var name_and_type_index = reader.read16();  
                classImage.constant_pool.push( {  tag: tag, bootstrap_method_attr_index: bootstrap_method_attr_index,name_and_type_index:name_and_type_index} );   
                break;
            default:                
                throw new Error(util.format("tag %s is not supported.", tag));
                //console.warn(util.format("tag %s is not supported.", tag));
        }
    }
        
    classImage.access_flags = reader.read16();
    
    classImage.this_class = reader.read16();
    
    classImage.super_class = reader.read16();

    
    classImage.interfaces = [];
    var interfaces_count = reader.read16();
    for(var i=0; i<interfaces_count; i++) {
        var index = reader.read16();
        if (index != 0){
            classImage.interfaces.push(index);
        }
    }
        
    classImage.fields = [];
    var fields_count = reader.read16();
    for(var i=0; i<fields_count; i++) {
        var access_flags = reader.read16();
        var name_index = reader.read16();
        var descriptor_index = reader.read16();
        var attributes_count = reader.read16();
        var field_info = {
            access_flags: access_flags,
            name_index: name_index,
            descriptor_index: descriptor_index,
            attributes_count: attributes_count,
            attributes: []
        }
        for(var j=0; j <attributes_count; j++) {
            var attribute_name_index = reader.read16();
            var attribute_length = reader.read32();
            var constantvalue_index = reader.read16();            
            var attribute = {
                attribute_name_index: attribute_name_index,
                attribute_length: attribute_length,
                constantvalue_index: constantvalue_index
            }            
            field_info.attributes.push(attribute);
        }
        classImage.fields.push(field_info);
    }    
    
    
    classImage.methods = [];
    var methods_count = reader.read16();
    for(var i=0; i<methods_count; i++) {
        var access_flags = reader.read16();
        var name_index = reader.read16();
        var signature_index = reader.read16();
        var attributes_count = reader.read16();
        var method_info  = {
            access_flags: access_flags,
            name_index: name_index,
            signature_index: signature_index,
            attributes_count: attributes_count,
            attributes: []
        }        
        for(var j=0; j <attributes_count; j++) {
            var attribute_name_index = reader.read16();
            var attribute_length = reader.read32();
            var info = getAttribues(attribute_name_index, reader.readBytes(attribute_length));            
            var attribute = {
                attribute_name_index: attribute_name_index,
                attribute_length: attribute_length,
                info: info
            }
            method_info.attributes.push(attribute);
        }
                
        classImage.methods.push(method_info);
    }
    
    
    classImage.attributes = [];
    var attributes_count = reader.read16();
    for(var i=0; i<attributes_count; i++) {
            var attribute_name_index = reader.read16();
            var attribute_length = reader.read32();
            var info = getAttribues(attribute_name_index, reader.readBytes(attribute_length));            
            var attribute = {
                attribute_name_index: attribute_name_index,
                attribute_length: attribute_length,
                info: info
            }
            classImage.attributes.push(attribute);        
    }
    
    return classImage;
 
};
